mlx_engine/stop_string_processor.py

```diff
@@ -0,0 +1,174 @@
+    stop_string: Optional[str] = None  # populated if status is "full_stop"
+    stop_tokens: Optional[List[int]] = None  # populated if status is "full_stop"
+
+
+class StopStringProcessor:
+    """State-fully processes tokens to check for stop strings during generation."""
+
+    def __init__(self, stop_strings: List[str], tokenizer):
+        """
+        Args:
+            stop_strings: List of strings that should stop generation if found
+            tokenizer: Tokenizer instance for encoding token IDs to text
+        """
+        # It is critical to calculate stop strings using both strings (for inter-token stop string detection)
+        # and token sequences (for multi-token strings). An example of a multi-token string for Llama-3.1 is
+        # "🌟" which is made up of the token sequence [9468, 234, 253]. If using only strings, partial matches
+        # can not be detected for multi-token strings.
+        self.stop_strings = stop_strings
+        self.stop_token_sequences = [
+            tokenizer.encode(stop_string) for stop_string in stop_strings
+        ]
+        self.tokenizer = tokenizer
+        self.token_id_buffer = []
+
+    def process_token(self, tokens: int) -> StopStringProcessorResult:
+        """Process a new token and check if generation should stop.
+
+        Args:
+            tokens: New token to process
+
+        Returns:
+            StopProcessorResult indicating the state of stop string detection
+        """
+        if len(self.stop_strings) == 0:
+            return StopStringProcessorResult(
+                status="no_match", stop_string=None, stop_tokens=None
+            )
+
+        self.token_id_buffer.append(tokens)
+
+        result = stopping_criteria(
+            token_sequence=self.token_id_buffer,
+            string=self.tokenizer.decode(self.token_id_buffer),
+            stop_token_sequences=self.stop_token_sequences,
+            stop_strings=self.stop_strings,
+        )
+
+        if result.status == "no_match":
+            # Can clear the buffer in no partial or full matches with stop sequences
+            self.token_id_buffer = []
+            return StopStringProcessorResult(
+                status="no_match", stop_string=None, stop_tokens=None
+            )
+
+        elif result.status == "partial_match":
+            return StopStringProcessorResult(
+                status="partial_match", stop_string=None, stop_tokens=None
+            )
+
+        elif result.status == "full_stop":
+            return StopStringProcessorResult(
+                status="full_stop",
+                stop_string=result.stop_string,
+                stop_tokens=self.token_id_buffer,
+            )
+
+        else:
+            raise ValueError(f"Unknown StopProcessorStatus: {result.status}")
+
+
+class StoppingCriteriaResult(NamedTuple):
+    status: StopStringProcessorStatus
+    stop_string: Optional[str] = None  # populated if status is "full_stop"
+
+
+def stopping_criteria(
```

Comment:
Can these methods be moved into the class and given a leading underscore to mark them as private?